BOOKKEEPER-1028 and BOOKKEEPER-1029

reddycharan opened the pull request:
BOOKKEEPER-1028: inc/excl opts listunderreplicated
    
    - Introduce including and excluding BookieId options
    for listunderreplicatedLedgers
    
    - But there is limitation that, since replicaslist wont be
    updated in underreplicatedledger zNode there is possibility
    of stale information

      ---------------------------------------------------------

    BOOKKEEPER-1029: BookieDecommision Workflow
    
    - LostBookieRecoveryDelay config param is stored in ZK
    - if LostBookieRecoveryDelay is reset to same value then it force triggers audit immediately
    - Added logic to trigger immediately or schedule pending audittask depending on the changed value in ZK
    - good number of testcases validating focetrigger/reschedluing audittask
    - added bookieshell command to get/set LostBookieRecoveryDelay from ZK
    - added bookieshell command to triggeraudit by resetting LostBookieRecoveryDelay
    - added decommissionbookie bkshell command, which validates the complete replication of ledgers stored in the bookie

reddycharan commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1848,6 +1909,288 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+            long firstEntryIter = firstEntryId;
+            // following the same approach followed in LedgerFragment.getFirstStoredEntryId()
+            for (int i = 0; i < ensembleSize && firstEntryIter <= lastEntryId; i++) {
+                if (distributionSchedule.hasEntry(lastEntryId, thisBookieIndexInCurrentEnsemble)) {
```
this has to be, 

hasEntry(firstEntryIter, thisBookieIndexInCurrentEnsemble) 

not

hasEntry(lastEntryId, thisBookieIndexInCurrentEnsemble)

eolivelli replied:
@reddycharan  can you fix this too ?

reddycharan replied:
fixed it

eolivelli commented:
@reddycharan I have started a review.

All tests run without failures on my laptop (after rebasing to current master)
Apache Rat check is OK 

But I see the findbugs errors below, can you fix them ?
Maybe you can rebase to current master, now the full test suite runs in 20 minutes

```
[INFO] Result of integer multiplication cast to long in org.apache.bookkeeper.bookie.BookieShell$DecommissionBookieCmd.waitForLedgersToBeReplicated(Collection, BookieSocketAddress, LedgerManager) [org.apache.bookkeeper.bookie.BookieShell$DecommissionBookieCmd] At BookieShell.java:[line 2081] ICAST_INTEGER_MULTIPLY_CAST_TO_LONG
[INFO] Boxing/unboxing to parse a primitive org.apache.bookkeeper.bookie.BookieShell$LostBookieRecoveryDelayCmd.runCmd(CommandLine) [org.apache.bookkeeper.bookie.BookieShell$LostBookieRecoveryDelayCmd] At BookieShell.java:[line 1403] DM_BOXED_PRIMITIVE_FOR_PARSING
[INFO] Boxing/unboxing to parse a primitive org.apache.bookkeeper.meta.ZkLedgerUnderreplicationManager.getLostBookieRecoveryDelay() [org.apache.bookkeeper.meta.ZkLedgerUnderreplicationManager] At ZkLedgerUnderreplicationManager.java:[line 741] DM_BOXED_PRIMITIVE_FOR_PARSING

```

sijie commented:
@reddycharan do you mind addressing the findbugs error?

reddycharan commented:
sure, will do. I was just waiting for full code review and comments, so that I can address all of them at once.

Thanks,
Charan

reddycharan commented:
@eolivelli @sijie 

fixed rat check warnings you mentioned. 
Yes 'decommissionbookie' bkshell command has to be executed in the node, which is being decommissioned.
added considerable number of comments and added clear description to the commit explaining about auditor trigger logic.

eolivelli commented:
I am sorry @reddycharan but findbugs is still failing

[INFO] Boxing/unboxing to parse a primitive org.apache.bookkeeper.bookie.BookieShell$LostBookieRecoveryDelayCmd.runCmd(CommandLine) [org.apache.bookkeeper.bookie.BookieShell$LostBookieRecoveryDelayCmd] At BookieShell.java:[line 1403] DM_BOXED_PRIMITIVE_FOR_PARSING

reddycharan commented:
@eolivelli I fixed it. It seems this time it has passed. Boxing/unboxing checks is new thing to me. Thanks for the follow up.

eolivelli commented:
jenkins job crashed, I restarted it
https://builds.apache.org/job/bookkeeper-precommit-pullrequest/54/console

CI completed with 2 tests failures, I think they are not releated to the patch

Failed tests:
TestSpeculativeRead.testSpeculativeReadFirstReadCompleteIsOk:260
TestBackwardCompat.testCompat410:647 Shouldn't be able to write

Tests run: 1167, Failures: 2, Errors: 0, Skipped: 0

reddycharan commented:
@sijie if you dont have anymore comments, this commit also should be good to go.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1337,6 +1354,68 @@ int runCmd(CommandLine cmdLine) throws Exception {
+            }
+            ZooKeeper zk = null;
+            try {
+                zk = ZooKeeperClient.newBuilder().connectString(bkConf.getZkServers())
```
It would be good to put this set/get logic in BookKeeperAdmin, for two seasons:

- it keeps the shell tool simpler
- it is easier to test and verify the logic.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1835,6 +1914,288 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+        @Override
+        public int runCmd(CommandLine cmdLine) throws Exception {
+            ZooKeeper zk = null;
+            LedgerUnderreplicationManager underreplicationManager = null;
```
same concerns above

It would be good to put this trigger logic in BookKeeperAdmin, for two seasons:

it keeps the shell tool simpler
it is easier to test and verify the logic.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1835,6 +1914,288 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+
+        @Override
+        public int runCmd(CommandLine cmdLine) throws Exception {
+            ZooKeeper zk = null;
```
you should put the decommision logic in bookkeeper admin. the shell here should be just a simple wrapper over that decommission method.

also we need tests for this behavior.

if you put the logic in Admin, you don't need to make DistributionSchedule public. These details should be hidden to the implementation.

reddycharan commented:
@sijie addressed your comments -
- move logic from BKShell to BookkeeperAdmin
- added functional testcases
- make DistributionSchedule and RoundRobinDistributionSchedule non-public

sijie commented:
@reddycharan there is a findbugs error introduced in this pull request. Can you fix that?

reddycharan commented:
@sijie  my bad, missed checking findbugs error after previous push. Fixed them now and it passed.

Thanks.

reddycharan commented:
@sijie @eolivelli fixed all findbug errors. this commit is good to go.

sijie commented:
Charan, +1. Do you want to try merge by yourself?
